design/external-names.md

```diff
@@ -0,0 +1,47 @@
+# External Names Service
+
+_Status_: Draft
+
+This document looks to allow Contour to process Kubernetes service types `ExternalName` so that traffic can be proxied to dns hosts rather than only pod endpoints.
+
+## Goals
+
+- Allow for Contour to proxy traffic via Envoy to external dns endpoints
+- Utilize built-in Kubernetes object types (i.e. ExternalName service)
+
+## Non-goals
+
+- `externalIPs` will not be used, users who want to proxy to custom IP's should build their own service/endpoint object
+- Utilize a new CRD or ConfigMap style mapping
+
+## Background
+
+Contour currently watches Kubernetes services and endpoints, streaming them to via Envoy xDS CDS & EDS endpoints.
+Requests to envoy are routed directly to Kubernetes pods (i.e. endpoints).
+This proposal looks to implement Kubernetes service type `ExternalName` which allows for an `externalName` which is represented as a valid RFC-1123 hostname.
+
+## High-level Design
+
+Contour will need to watch for and process Kubernetes service types of `ExternalName` (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#servicespec-v1-core).
+When configuring EDS endpoints, for this service type Contour will pass the contents of the `externalName` field on the Kubernetes service to EDS instead of a PodIP.
```

Comment:
I'm not sure this is correct. Does the endpoint controller create an Endpoint document for services that use ExternalName? If there is no endpoint there will be no entry in EDS.

What we'll do instead is record the DNS name as a new kind of dag object (see below for a suggestion) and the cluster visitor will configure the cluster discovery type to use DNS (https://www.envoyproxy.io/docs/envoy/v1.9.0/api-v2/api/v2/cds.proto#envoy-api-enum-cluster-discoverytype) rather than EDS.

Reply:
Oh good catch, when I started this, in my head I was going to send a dns entry to EDS instead of a IP, but later found the other way which I referenced later. Let me update.